extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -0,0 +1,371 @@
+            resources./*OK*/changeSize(element, undefined, attributeValue);
+          } else if (property === 'height') {
+            user().assert(typeof attributeValue === 'number',
+                'Invalid result for [height]: %s', attributeValue);
+            resources./*OK*/changeSize(element, attributeValue, undefined);
+          }
+          element.attributeChangedCallback(property, oldValue, attributeValue);
+        }
+      }
+    }
+  }
+
+  /**
+   * If the current value of `binding` equals `expectedValue`, returns true.
+   * Otherwise, returns false.
+   * @param {!BindingDef} binding
+   * @param {BindExpressionResultDef} expectedValue
+   * @private
+   */
+  verifyBinding_(binding, expectedValue) {
+    const property = binding.property;
+    const element = binding.element;
+
+    // TODO(choumx): Support objects for attributes.
+
+    let initialValue;
+    let match = true;
+
+    if (property === 'text') {
+      initialValue = element.textContent;
+      match = (initialValue.trim() === expectedValue.trim());
+    } else if (property === 'class') {
+      initialValue = element.classList;
+
+      /** @type {!Array<string>} */
+      let classes = [];
+      if (Array.isArray(expectedValue)) {
+        classes = expectedValue;
+      } else if (typeof expectedValue === 'string') {
+        classes = expectedValue.split(' ');
+      } else {
+        user().error(TAG,
+            'Unsupported result for class binding', expectedValue);
+      }
+      match = this.compareStringArrays_(initialValue, classes);
+    } else {
+      const attribute = element.getAttribute(property);
+      if (typeof expectedValue === 'boolean') {
+        initialValue = !!attribute;
+      } else {
+        initialValue = attribute;
+      }
+      // Use abstract equality since boolean attributes return value of ''.
+      match = (initialValue == expectedValue);
+    }
+
+    if (!match) {
+      user().error(TAG,
+        `<${element.tagName}> element [${property}] binding ` +
+        `default value (${initialValue}) does not match first expression ` +
+        `result (${expectedValue}).`);
+    }
+  }
+
+  /**
+   * Returns true if both arrays contain the same strings.
+   * @param {!(IArrayLike<string>|Array<string>)} a
+   * @param {!(IArrayLike<string>|Array<string>)} b
+   * @return {boolean}
+   */
+  compareStringArrays_(a, b) {
+    if (a.length !== b.length) {
+      return false;
+    }
+    const sortedA = a.slice().sort();
+    const sortedB = b.slice().sort();
+    for (let i = 0; i < a.length; i++) {
+      if (sortedA[i] !== sortedB[i]) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Returns `value` if it's an appropriate Element attribute type.
+   * Otherwise, returns null.
+   * @param {BindExpressionResultDef} value
+   * @return {(string|boolean|number|null)}
+   */
+  attributeValueOf_(value) {
+    if (typeof value === 'string' || typeof value === 'boolean'
```

Comment:
I actually had it that way originally but couldn't get it to work with Closure's type checker. I'll give it another try.

Also changed to use a hash as a minor optimization.